rust-symcrypt/src/rsa.rs

```diff
@@ -0,0 +1,643 @@
+                return Err(result.into());
+            }
+        }
+
+        let pub_exp_bytes = store_msb_first_u64(pub_exp[0], self.size_of_public_exponent())?;
+
+        Ok(RsaKeyPairBlob {
+            modulus: modulus_buffer,
+            pub_exp: pub_exp_bytes,
+            p,
+            q,
+            d_p,
+            d_q,
+            crt_coefficient,
+            private_exp: private_exponent,
+        })
+    }
+
+    /// `export_public_key_blob()` will export a [`RsaPublicKey`].
+    pub fn export_public_key_blob(&self) -> Result<RsaPublicKeyBlob, SymCryptError> {
+        let mut modulus_buffer = vec![0u8; self.size_of_modulus() as usize];
+        let mut pub_exp = vec![0u64; 1];
+        unsafe {
+            // SAFETY: FFI calls
+            // When only getting the public key, ppPrimes, pcbPrimes and nPrimes can be NULL, NULL and 0.
+            match symcrypt_sys::SymCryptRsakeyGetValue(
+                self.inner,
+                modulus_buffer.as_mut_ptr(),
+                modulus_buffer.len() as symcrypt_sys::SIZE_T,
+                pub_exp.as_mut_ptr(),
+                1 as symcrypt_sys::UINT32,
+                ptr::null_mut(),
+                ptr::null_mut(),
+                0,
+                NumberFormat::MSB.to_num_format(),
+                self.key_usage.to_flag(),
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(RsaPublicKeyBlob {
+                    modulus: modulus_buffer,
+                    pub_exp: store_msb_first_u64(pub_exp[0], self.size_of_public_exponent())?,
+                }),
+                err => Err(err.into()),
+            }
+        }
+    }
+
+    /// `size_of_modulus()` returns a `u32` representing the (tight) size in bytes of a byte array big enough to store
+    /// the modulus of the key.
+    pub fn size_of_modulus(&self) -> u32 {
+        unsafe {
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptRsakeySizeofModulus(self.inner)
+        }
+    }
+
+    /// `size_of_public_exponent()` returns a `u32` representing the (tight) size in bytes of a byte array big enough to store
+    /// the public exponent of the key.
+    pub fn size_of_public_exponent(&self) -> u32 {
+        unsafe {
+            // SAFETY: FFI calls
+            // Only one public exponent is supported, so the only valid index is 0.
+            symcrypt_sys::SymCryptRsakeySizeofPublicExponent(self.inner, 0)
+        }
+    }
+
+    /// `key_usage` returns the intended usage for the RSA key pair.
+    pub fn key_usage(&self) -> RsaKeyUsage {
+        self.key_usage
+    }
+
+    /// `key_usage` returns the intended usage for the RSA key pair.
+    pub(crate) fn inner(&self) -> symcrypt_sys::PSYMCRYPT_RSAKEY  {
+        self.inner
+    }
+}
+
+impl Drop for RsaKeyPair {
+    fn drop(&mut self) {
+        unsafe {
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptRsakeyFree(self.inner);
+        }
+    }
+}
+
+impl RsaPublicKey {
```

Comment:
>RsaPublicKey [](http://example.com/codeflow?start=5&length=12)

Interesting architectural decision here.

You chose to have an RsaKey object and an RsaPublicKey object. SymCrypt uses just an RsaKey object, and sometimes that contains only a public key and sometimes it contains a full key pair. 

Both solutions work, just make sure you're fitting into the style of other Rust APIs.

Reply:
I left a comment about this above; I'm not in favor of this solution